Bootstrap await utility

siteshJalan opened the pull request:
## Overview

Create await utility with supported tests for testing the bootup of various resources. 

## Details

The resources which will will be tested as part of this utility will be.

- http
- redis
- mongodb
- dynamodb
- memcached
- mysql

## References

- [SRE-5](https://saltside.atlassian.net/browse/SRE-5)

## Todo

- Basic functionality
  * [x] Http check
  * [x] redis check
  * [x] dynamodb check
  * [x] Mysql check
  * [x] memcached check
- UX
  * [x] Drop timeout in favor of single `-r` (for retry) option
  * [ ] Add usage/help message on incorrect invocation
  * [ ] Add `await help`
  * [x] Support `-- [EXTRA_OPTIONS]` style for every check
- [ ] Write readme. Include a how to develop section. This should be targeted at all contributors (and not just saltside)

siteshJalan committed:
Add await utility

Add await utility with supported tests for testing http, redis and
mongodb connections.

SRE-5

siteshJalan committed:
Add dynamodb to the test utility

ahawkins commented on bin/await:
```diff
@@ -0,0 +1,174 @@
+			shift $((OPTIND-1))
+
+			while [ "${counter}" -lt "${retry}" ]; do
+				if mongo --quiet "${host}:${port}/local" --eval 'db.stats()' > /dev/null ; then
```
I think we should drop the host and port arguments and use `-- [MONGO_OPTS`. This way people can also specify user names, passwords, or whatever other junk is required to connect a general mongo server.

ahawkins commented on bin/await:
```diff
@@ -0,0 +1,174 @@
+			shift $((OPTIND-1))
+
+			while [ "${counter}" -lt "${retry}" ]; do
+				if curl -s --fail -m "${timeout}" "${host}" > /dev/null; then
```
We should include `$@` here as well. This will allow people to set headers and whatever other stuff they like.

ahawkins commented on bin/await:
```diff
@@ -0,0 +1,174 @@
+			local host timeout=60 counter=0 retry=3
+
+			while getopts ":h:t:r:" opt; do
+
```
You can kill this blank line

ahawkins commented on bin/await:
```diff
@@ -0,0 +1,174 @@
+	local subcommand="${1:-}"
+	shift
+
+	case "$subcommand" in
```
Please use `${subcommand}`.

ahawkins commented on bin/await:
```diff
@@ -0,0 +1,174 @@
+			;;
+
+		mongodb)
+			local host port=27017 counter=0 retry=3
```
Please add a blank line

ahawkins commented on bin/await:
```diff
@@ -0,0 +1,174 @@
+			;;
+
+		redis)
+			local host port=6379 counter=0 retry=3
```
please add a blank line

ahawkins commented on docker-compose.yml:
```diff
@@ -0,0 +1,22 @@
+    build: .
+    environment:
+      DYNAMO_DB_URL: http://dynamodb:8000
+      AWS_ACCESS_KEY_ID : "test"
```
IMO we should remove these aws key variables and pass them to the test. This way we ensure the command passes `$@` along to `aws`.

siteshJalan replied:
adam these variables need to be present in environment. I cannot find an option in aws-cli where i can pass it directly. Only thing i find is via using profiles. Do you know the options which i can use?

ahawkins reviewed (changes requested):
Functionally this looks like it will as expected. We can improve the UX though. I suggest we drop the concept for retries and timeouts. Instead we should have a global `-t` (for timeout option) in seconds. This works for the looping structure. Specific command options such as curl opts should be used to handle those cases.

ahawkins commented:
@siteshJalan can we add memcached as well?

ahawkins committed:
Reduce image size by using apline linux

ahawkins committed:
Extract virtual buildeps apk set

terlar committed:
Add TravisCI configuration

terlar committed:
Add build status to README

terlar committed:
Build and push docker image to Docker Hub

terlar committed:
Use official tagging pattern

This introduces a tagging pattern with major, minor and patch version.

Commit:
Merge pull request #3 from Saltside/task/connect-ci

Connect to CI

siteshJalan commented:
@ahawkins : may i drop `-t` in favour of `$@` using which a caller can specifically pass any timeouts needed. We rather give just `-r` which will be used as retry count? what do you think?

siteshJalan committed:
Add mysql to await utility

This commit adds mysql connection chck to await utility. Along with it
it also removes the support of `-t` from various mudules. Now they
should be using just `-r` to specift the number of retries. Any timeout
should be passed a command argument i.e. after end of options denoted by
`--`.

terlar committed:
Replace	mongodb dependency with python client

Since the mongodb adds over 100MB (because of the server part), it
increases the image with more than double. To avoid this we use a custom
wrapper around the python client.

terlar committed:
Support extra options for mongostats

Commit:
Merge pull request #2 from Saltside/feature/shrink

Extract virtual buildeps apk set

siteshJalan committed:
Add memcached to await utility

ahawkins commented on test/await_test.bats:
```diff
@@ -0,0 +1,76 @@
+}
+
+@test "successful mysql connection with retry" {
+	run await mysql -h mysql -r 5 -- --user=root --password=secret mydb
```
These should be environment variables.

siteshJalan replied:
thought of putting it here so that `$@` could be tested as well.

terlar committed:
Streamline mongo check

- Use seconds for timeout
- Use url instead of uri parameter name
- Rename to mongo_check instead of mongostats
- Don't output the stats, but just output a simple OK!

terlar committed:
Rebuild docker image when dependencies change

terlar committed:
Use python lib for mysql check

The regular MySQL client requires you to specify a database, but since
some checks might not have a database when you want to check if the
server is up we cannot rely on that.

terlar committed:
Use python lib for redis check

terlar committed:
Use full URL instead of host for http

terlar committed:
Unify style for semicolon after command

terlar committed:
Use python lib for dynamodb check

terlar committed:
Use python lib for memcached check

terlar committed:
Unify order of parameters

ahawkins commented:
@terlar this may be bikeshedding, but I have a proposal for simplify the UX given you've updated all checks to use URIs. Here's my proposal:

```
$ await [ -r RETRY_COUNT] URI -- [ CONNECTION_OTPS ]

# Real world eamples
$ await [ -r RETRY_COUNT] redis://host:port/db -- [ REDIS_CONNECTION_OPTS ]
$ await [ -r RETRY_COUNT] mysql://user:password@host:port -- [ MYSQL_CONNECTION_OPTS ]
$ await [ -r RETRY_COUNT] dynamodb://host:port -- [ DYNAMODB_CONNECTION_OPTS ]
$ await [ -r RETRY_COUNT] http(s)://host:port/path -- [ HTTP_CONNECTION_OPTS ]
$ await [ -r RETRY_COUNT] memcached://host:port -- [ MEMCACHED_CONNECTION_OPTS]
```

Essentially we can drop the subcommand and do everything via the protocol in the URI argument. Thoughts?

terlar commented:
Yeah, that looks cleaner, I can't come up with any problem doing that at this point.

terlar commented:
The only tricky part is with the memcached, I think they use tcp:// prefix? Also dynamodb use http:// as prefix.

ahawkins commented:
> The only tricky part is with the memcached, I think they use tcp:// prefix? Also dynamodb use http:// as prefix.

I suggest we use the protocol in the first positional argument to determine the check type (e.g. http, mysql, memcached etc). Then in the handler for the check the type can be replaced with something specific if needed. It depends also on _how_ we do the check for each one. What do you think about this:

- `http(s)://` - curl
- `mongo://` - python script
- `mysql://` - python script
- `dynamodb://` - python script
- `memcached` - python script
- `redis` - python script

Where possible we can just handle the full URI off to a library else we can split it ourselves.

ahawkins committed:
WIP: rework to only use URIs

ahawkins commented:
@terlar I took a spike on this in https://github.com/Saltside/await/pull/4. Look at all those deletions :D

ahawkins committed:
Make tests pass

This commit also make the dynamodb check work against local DynamoDB
only. It sets placeholder values for region and AWS keys.

ahawkins committed:
Add generic "await cmd -- [CMD]" form

ahawkins committed:
Add usage checks

ahawkins committed:
Reduce connection timeouts from 3 to 1 seconds

ahawkins committed:
Only push images on master builds

ahawkins committed:
Add aws CLI to image

ahawkins committed:
Comment curious -- test

Commit:
Merge pull request #4 from Saltside/feature/uris

rework to only use URIs
